### PR TITLE
Prevent use of school email address

### PR DESCRIPTION
**Pull request by hortha**
### Context

Ideally DfE identity accounts should be associated with personal email addresses so teachers can keep the same account as they progress throughout their career in different education establishments

### Changes proposed in this pull request

Add a new table establishment_domains to store a list of school/establishment domains that should NOT be used for email addresses.
Add a job to refresh the list of domains using the school website field from the GAIS download.
Add validation of the email suffix in a number of pages using the data in this new table.

### Guidance to review

Model, Background Job, Page and test changes

### Checklist

-   [x] Attach to Trello card
-   [x] Rebased master
-   [ ] Cleaned commit history
-   [x] Tested by running locally
-   [ ] Reminder created to manually clean any removed app settings post deployment

### CODE REVIEW

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Jobs/RefreshEstablishmentDomainsJob.cs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+using Microsoft.EntityFrameworkCore;
+using TeacherIdentity.AuthServer;
+using TeacherIdentity.AuthServer.Models;
+using TeacherIdentity.AuthServer.Services.Establishment;
+
+namespace TeacherIdentity.AuthServer.Jobs;
+
+public class RefreshEstablishmentDomainsJob
+{
+    private const string ExtractDomainRegex = "^(?:https?:\\/\\/)?(?:[^@\\/\\n]+@)?(?:www\\.)?([^:\\/\\n]+)";
```
Comment (gunndabad) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Jobs/RefreshEstablishmentDomainsJob.cs:12]: May a little more efficient to stash a `Regex` instance here and use source generator e.g.
```
[GeneratedRegex("^(?:https?:\\/\\/)?(?:[^@\\/\\n]+@)?(?:www\\.)?([^:\\/\\n]+)")]
private static partial Regex MatchDomainRegex();
```

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Models/Mappings/EstablishmentDomainMapping.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
+
+namespace TeacherIdentity.AuthServer.Models.Mappings;
+
+public class EstablishmentDomainMapping : IEntityTypeConfiguration<EstablishmentDomain>
+{
+    public void Configure(EntityTypeBuilder<EstablishmentDomain> builder)
+    {
+        builder.ToTable("establishment_domains");
```
Comment (gunndabad) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Models/Mappings/EstablishmentDomainMapping.cs:10]: Worth throwing on the case insensitive collation here?

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Jobs/RefreshEstablishmentDomainsJob.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        {
+            await SaveChangesAndLog(cancellationToken);
+        }
+
+        sw.Stop();
+        _logger.LogInformation($"Refresh establishment domain job completed in {sw.ElapsedMilliseconds}ms.");
+
+        bool TryAddEstablishmentDomain(string domainName)
+        {
+            if (existingDomains!.Contains(domainName))
```
Comment (gunndabad) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Jobs/RefreshEstablishmentDomainsJob.cs:82]:
```suggestion
            if (existingDomains!.Contains(domainName, StringComparer.OrdinalIgnoreCase))
```

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Services/Establishment/CsvDownloadEstablishmentMasterDataService.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    public CsvDownloadEstablishmentMasterDataService(
+        HttpClient httpClient,
+        IClock clock)
+    {
+        _httpClient = httpClient;
+        _clock = clock;
+    }
+
+    public async IAsyncEnumerable<string?> GetEstablishmentWebsites()
```
Comment (gunndabad) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Services/Establishment/CsvDownloadEstablishmentMasterDataService.cs:20]: Nice 🚀

**Review by gunndabad — APPROVED** [review]

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Jobs/RefreshEstablishmentDomainsJob.cs (hunk 1 above)
Comment (hortha) [reply dotnet-authserver/src/TeacherIdentity.AuthServer/Jobs/RefreshEstablishmentDomainsJob.cs]: Nice! - done

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Jobs/RefreshEstablishmentDomainsJob.cs (hunk 3 above)
Comment (hortha) [reply dotnet-authserver/src/TeacherIdentity.AuthServer/Jobs/RefreshEstablishmentDomainsJob.cs]: Done

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Models/Mappings/EstablishmentDomainMapping.cs (hunk 2 above)
Comment (hortha) [reply dotnet-authserver/src/TeacherIdentity.AuthServer/Models/Mappings/EstablishmentDomainMapping.cs:10]: Yep - done

**Pull request merged**